File: 0079-Word-Search/0079-Word-Search.go
func helper(board [][]byte, i, j, index int, word string) bool {
	if index == len(word) {
		return true
	}
	if i < 0 || i >= len(board) || j < 0 || j >= len(board[0]) ||
		board[i][j] == '*' || board[i][j] != word[index] {
		return false
	}
	temp := board[i][j]
	board[i][j] = '*'
	if helper(board, i+1, j, index+1, word) ||
		helper(board, i-1, j, index+1, word) ||
		helper(board, i, j+1, index+1, word) ||
		helper(board, i, j-1, index+1, word) {
		return true
	}
	board[i][j] = temp
	return false
}

func exist(board [][]byte, word string) bool {
	for i := 0; i < len(board); i++ {
		for j := 0; j < len(board[0]); j++ {
			if board[i][j] == word[0] {
				if helper(board, i, j, 0, word) {
					return true
				}
			}
		}
	}
	return false
}

File: 0079-Word-Search/0079-Word-Search.java
class Solution {
    boolean helper(char[][] board, int i, int j, String word, int index) {
        if (index == word.length())
            return true;
        if (i < 0 || i > board.length - 1 || j < 0 || j > board[0].length - 1 ||
                board[i][j] == '*' || board[i][j] != word.charAt(index)) {
            return false;
        }
        char temp = board[i][j];
        board[i][j] = '*';
        if (helper(board, i + 1, j, word, index + 1) ||
            helper(board, i - 1, j, word, index + 1) ||
            helper(board, i, j + 1, word, index + 1) ||
            helper(board, i, j - 1, word, index + 1)) {
                return true;
        }

        board[i][j] = temp;
        return false;
    }

    public boolean exist(char[][] board, String word) {
        boolean result;
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                if (board[i][j] == word.charAt(0)) {
                    result = helper(board, i, j, word, 0);
                    if (result)
                        return true;
                }
            }
        }
        return false;
    }
}
